src/Particular.Analyzers/Cancellation/CatchAllShouldOmitOperationCanceledAnalyzer.cs

```diff
@@ -0,0 +1,133 @@
+            {
+                var catchType = GetCatchType(catchClause);
+
+                if (catchType == "OperationCanceledException")
+                {
+                    return;
+                }
+
+                if (catchType == "Exception" || catchType == "System.Exception")
+                {
+                    if (CatchFiltersOutOperationCanceled(catchClause, context))
+                    {
+                        return;
+                    }
+
+                    var tryBlockCalls = tryStatement.ChildNodes().OfType<BlockSyntax>()
+                        .SelectMany(block => block.DescendantNodes().OfType<InvocationExpressionSyntax>());
+
+                    foreach (var call in tryBlockCalls)
+                    {
+                        if (call.ArgumentList.Arguments.Any(arg => IsCancellationToken(arg.Expression, context, cancellationTokenType)))
+                        {
+                            context.ReportDiagnostic(DiagnosticDescriptors.CatchAllShouldOmitOperationCanceled, catchClause.CatchKeyword);
+                            return;
+                        }
+                    }
+
+                }
+            }
+        }
+
+        static bool CatchFiltersOutOperationCanceled(CatchClauseSyntax catchClause, SyntaxNodeAnalysisContext context)
+        {
+            var filterClause = catchClause.ChildNodes().OfType<CatchFilterClauseSyntax>().FirstOrDefault();
+            if (filterClause == null)
+            {
+                return false;
+            }
+
+            var logicalNotExpression = filterClause.ChildNodes().OfType<PrefixUnaryExpressionSyntax>().FirstOrDefault();
+            if (logicalNotExpression == null || !logicalNotExpression.IsKind(SyntaxKind.LogicalNotExpression))
+            {
+                return false;
+            }
+
+            var parenthesizedExpression = logicalNotExpression.ChildNodes().OfType<ParenthesizedExpressionSyntax>().FirstOrDefault();
+            if (parenthesizedExpression == null)
+            {
+                return false;
+            }
+
+            var binaryExpression = parenthesizedExpression.ChildNodes().OfType<BinaryExpressionSyntax>().FirstOrDefault();
+            if (binaryExpression == null)
+            {
+                return false;
+            }
+
+            var isKeyword = binaryExpression.OperatorToken.IsKind(SyntaxKind.IsKeyword);
+            var leftSymbol = context.SemanticModel.GetSymbolInfo(binaryExpression.Left, context.CancellationToken).Symbol as ILocalSymbol;
+            var rightSymbol = context.SemanticModel.GetSymbolInfo(binaryExpression.Right, context.CancellationToken).Symbol as INamedTypeSymbol;
+
+            var leftIsException = leftSymbol?.Type.ToString() == "System.Exception";
+            var rightIsOpCanceled = rightSymbol?.ToString() == "System.OperationCanceledException";
+
+            return isKeyword && leftIsException && rightIsOpCanceled;
```

Comment:
We can bail out before each hit of the semantic model here:

```suggestion
            if (!binaryExpression.OperatorToken.IsKind(SyntaxKind.IsKeyword))
            {
                return false;
            }

            var leftSymbol = context.SemanticModel.GetSymbolInfo(binaryExpression.Left, context.CancellationToken).Symbol as ILocalSymbol;

            if (leftSymbol?.Type.ToString() != "System.Exception")
            {
                return false;
            }

            var rightSymbol = context.SemanticModel.GetSymbolInfo(binaryExpression.Right, context.CancellationToken).Symbol as INamedTypeSymbol;

            return rightSymbol?.ToString() == "System.OperationCanceledException";
```